pypeit/images/combineimage.py

```diff
@@ -149,16 +152,24 @@ def run(self, bias=None, flatimages=None, ignore_saturation=False, sigma_clip=Tr
         img_list = [img_stack]
         var_stack = utils.inverse(ivar_stack)
         var_list = [var_stack, rn2img_stack]
-        img_list_out, var_list_out, outmask, nused = combine.weighted_combine(
-            weights, img_list, var_list, (mask_stack == 0),
-            sigma_clip=sigma_clip, sigma_clip_stack=img_stack, sigrej=sigrej, maxiters=maxiters)
+        if combine_method == 'weightmean':
+            img_list_out, var_list_out, gpm, nused = combine.weighted_combine(
+                weights, img_list, var_list, (mask_stack == 0),
+                sigma_clip=sigma_clip, sigma_clip_stack=img_stack, sigrej=sigrej, maxiters=maxiters)
+        elif combine_method == 'median':
```

Comment:
My original reason to moving to `cython` (way back in the day) was that the masked median operation in numpy had huge (unacceptable) overhead. This seems fine to just do a median... but I wonder if we can find a clever way to do a masked median. Here's one possible idea (**note:** I just wrote this out, and haven't checked it, but I think it's OK...). Hope you get the general idea:

```
# Mask values with huge positive number, then sort it
img_stack[mask_stack==1] = 1E30
sorted_img_stack = np.sort(img_stack)
# Calculate how many masked values there are
nummsk = np.sum(mask_stack, axis=0)
# Deal with odd/even elements separately
idx, oddeven = np.divmod(nummsk, 2)
# Initialise the output image
final_image=np.zeros(nummsk.shape)
# Fill odd values
ww = np.where(oddeven == 1)
fw = (idx[ww], ww[0], ww[1])
final_image[ww] = sorted_img_stack[fw]
# Fill even values
ww = np.where((oddeven == 0) & (nummsk != img_stack.shape[0]))  # second term deals with cases where one pixel in the entire stack is masked
fm = (idx[ww]-1, ww[0], ww[1])
fp = (idx[ww], ww[0], ww[1])
final_image[ww] = 0.5*(sorted_img_stack[fm]+sorted_img_stack[fp])
img_list_out = [final_image]
```
Probably all of the above could be put into a neater `def`...

Reply:
This looks good to me. You need to update the gpm that is returned to have masked values at locations where all the values are basically masked. Also, I would avoid using 1 and 0 and == 0 and == 1 for booleans. I also try to use np.where as little as possible, but it looks like you need it here to slice x and y differently?

Reply:
One needs to project the indices from a 2D array to two elements of a 3D array, so I fear `np.where` is needed in this case... I haven't checked this code though... so if you can suggest a better alternative, I've all for it!!

Reply:
raised an Issue